Fixes #390, Fixes #427, Fixes #538, Fixes #613 Card Widget Updates

tadean opened the pull request:
This pull request updates many aspects of the card widget of our custom card field, primarily the introduction of card previews and two-step editor forms so that the entire deck does not open at once.

## Description
The following changes are made to the card type and card widgets:

- Storage of the media ID is sanitized so that it may be NULL
- Corrected issue where the media widget was providing a comma delimited list of Media IDs
- Card field forms are now a two-step AJAX process where an `Edit Card` button is pressed to edit an individual card, as opposed to all field inputs showing simultaneously.
- Card background color selection
- Card background color implementation
- On the edit screen, a rendered preview is provided for collapsed cards.

## Related Issue
#538 
#427 
#390 
#613 
#377 

## How Has This Been Tested?

- Create a page
- Add cards
- Verify that for new cards they do not initially appear collapsed (eg. all fields are accessible for cards newly added to the entity)
- Save page
- Edit page
- Edit cards paragraph
- Verify that each card you added has a visible preview, but that the form fields are hidden
- Click `Edit Card` and verify it expands to make its form fields accessible.
- Make changes
- Verify that an image can be removed from the card (Fixes #538)
- Click `Collapse Card` and verify it reverts back to preview mode with changes reflected. (or skip this step and leave it expanded)
- Save page
- Verify changes were updated
- Verify background colors can be set for cards
- Verify background colors appear for cards.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

tadean commented:
![image](https://user-images.githubusercontent.com/51835841/113359247-40b5c400-92fc-11eb-966a-5d7560737df3.png)

![image](https://user-images.githubusercontent.com/51835841/113359261-46aba500-92fc-11eb-9c12-f00dc87f8749.png)

![image](https://user-images.githubusercontent.com/51835841/113359278-4ca18600-92fc-11eb-972b-2745889de0d7.png)

tadean commented:
The following issue still affects cards which has pending resolution in Drupal core. It is not covered by this PR:

- [Allow for deletion of a single value of a multiple value field - General Field API issue](https://www.drupal.org/project/drupal/issues/1038316)

smayden reviewed (approved):
Tested as a content editor and verified that all the improvements outlined for this PR is working for as expected. Super impressed with the QS 2 implementation for cards!

I'll make a few notes as a separate comment for possible improvements down the line but none of them are really in the scope of this PR.

smayden commented:
A few things I noticed while testing this PR that might be improvements for cards down the line (not necessarily for this ticket):
-When I click edit on cards (actually all paragraphs) it seems to jump me down to the bottom of the edit area. I think this will confuse content editors
-We might consider making the edit/collapse buttons for each card in the upper right-hand corner to be consistent with where these buttons appear elsewhere on the edit screen.
-Is there a way to make the whole card clickable? I couldn't figure out how to do that (maybe we decided not to do that?)
-Should we consider being able to change the color of the link button? I made a card with red background and button blends with background (could button switch to reverse similar to how text does based on color?)
-Is there a reason we default to basic html rather than the standard text format? (I probably missed or forgot the conversation...) I was trying to make a card with my own button in the card body and was unable to do so until I switched to standard (and agreed to the scary warning).

danahertzberg commented:
Follow up tasks are documented here: https://github.com/az-digital/az_quickstart/issues/377

danahertzberg commented:
@tadean Will make easy updates from #377 before merging this.

geijutsuka reviewed (commented):
Works for me :)
There was an error when entering after uploading the media, but appears to be a Media/Core problem? This is all I could find relating to the error:
- https://www.drupal.org/forum/general/general-discussion/2020-10-20/the-specified-ajax-callback-is-empty-or-not-callable
- https://www.drupal.org/project/drupal/issues/2866330
Error:
`Symfony\Component\HttpKernel\Exception\HttpException: The specified #ajax callback is empty or not callable. in Drupal\Core\Form\FormAjaxResponseBuilder->buildResponse() (line 67 of /src/az-quickstart-scaffolding/web/core/lib/Drupal/Core/Form/FormAjaxResponseBuilder.php).`

joeparsons commented:
@tadean I'm seeing a strange issue when editing an existing card.  It seems when I edit an existing cards paragraph on an existing page, when the collapsed version of the card is rendered in its preview mode, the wrong image is used.  Here is a short screencast of the behavior on the Probo site:

https://user-images.githubusercontent.com/471936/113449802-78cc0e00-93b3-11eb-867e-bb33debf92b5.mp4

tadean commented:
> @tadean I'm seeing a strange issue when editing an existing card. It seems when I edit an existing cards paragraph on an existing page, when the collapsed version of the card is rendered in its preview mode, the wrong image is used. Here is a short screencast of the behavior on the Probo site:
> 
>  card-image-issue.mp4

Good find @joeparsons ! I suspect there's probably something we need to do in order to tell the renderer that the card image needs to be regenerated. Most fields probably handle that automatically and it may be something our compound field needs to do too. I'll check into that.

joeparsons commented:
Also, for what it's worth, I tried going through the steps of adding a card with Javascript disabled in my browser and found that clicking the "Add media" button for the media form element doesn't work _at all_.  I imagine that is proably an issue with the Media Library Form Element module we're using 😞.

tadean commented:
> Also, for what it's worth, I tried going through the steps of adding a card with Javascript disabled in my browser and found that clicking the "Add media" button for the media form element doesn't work _at all_. I imagine that is proably an issue with the Media Library Form Element module we're using .

Bummer :frowning_face: 

Maybe when that FAPI media library widget gets incorporated into core it'll be a little more complete. I don't think a modal could work without javascript, but there's no reason they couldn't make it a multistep form element.

joeparsons commented:
> Maybe when that FAPI media library widget gets incorporated into core it'll be a little more complete. I don't think a modal could work without javascript, but there's no reason they couldn't make it a multistep form element.

Actually it looks like it could be a limitation of the media_library module in core:
https://www.drupal.org/project/drupal/issues/2987675

joeparsons commented on modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:
```diff
@@ -20,11 +23,149 @@
+        '#type' => 'container',
+        '#attributes' => [
+          'class' => ['col-12', 'col-sm-6', 'col-md-6', 'col-lg-4', 'card-preview',
+          ],
```
Nitpick: is this closing bracket intentionally on a new line?

tadean replied:
I'll take a look at tweaking that - that one required some tinkering to get something `phpcs` was happy with. I do think it'd look better on the same line too.

tadean replied:
I think this should be resolved now.

trackleft commented on modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:
```diff
@@ -20,11 +23,149 @@
+   */
+  public function form(FieldItemListInterface $items, array &$form, FormStateInterface $form_state, $get_delta = NULL) {
+
+    // Create shared settings for widget elemeents.
```
```suggestion
    // Create shared settings for widget elements.
```

joeparsons commented:
> Good find @joeparsons ! I suspect there's probably something we need to do in order to tell the renderer's caching that the card image needs to be regenerated. Most fields probably handle that automatically and it may be something our compound field needs to do too. I'll check into that.

FWIW, it looks like the `AZCardDefaultFormatter::generateImageRenderArray()` method is already doing something to add cache dependencies via the renderer.  Perhaps that method could be moved into a service that could be used by both the Formatter and Widget plugins?

trackleft commented on modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:
```diff
@@ -20,11 +23,149 @@
+  public function form(FieldItemListInterface $items, array &$form, FormStateInterface $form_state, $get_delta = NULL) {
+
+    // Create shared settings for widget elemeents.
+    // This is necessary because wigets have to be AJAX replaced together,
```
```suggestion
    // This is necessary because widgets have to be AJAX replaced together,
```

trackleft commented:
> > Good find @joeparsons ! I suspect there's probably something we need to do in order to tell the renderer's caching that the card image needs to be regenerated. Most fields probably handle that automatically and it may be something our compound field needs to do too. I'll check into that.
> 
> FWIW, it looks like the `AZCardDefaultFormatter::generateImageRenderArray()` method is already doing something to add cache dependencies via the renderer. Perhaps that method could be moved into a service that could be used by both the Formatter and Widget plugins?


Some more context that may or may not be relevant: image only appears when card is collapsed within the edit form, also, the image that does appear is a temporary image https://take.ms/YZLbC

trackleft commented:
Could be related to that image that Rebecca received an error about

tadean commented:
Updated the following:

- Fixed some documentation typos (thanks @trackleft !)
- Reformatted an array initialization to be more readable (thanks @joeparsons !)
- Created class constant for default text format.

From #377 
- Use standard format for body by default
- Move background color selection to top
- Move media subfield to second top
- Change link_uri label to Card Link URL

tadean commented:
Pending tasks:

- [ ] Refactor card image generation to service

tadean commented:
Refactored to use a shared service between the formatter and widget to provide card image rendering.

joeparsons reviewed (approved):
All the recent changes look good to me.  I no longer see the issue with the wrong card image being rendered in the preview mode like I was before.  Nice work @tadean!

barrettbaffert reviewed (approved):
This is great! Very enjoyable editing experience. 

https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-634.probo.build/barrett-card-test

🃏🃏🃏

smayden reviewed (approved):
Tested as content editor and all working as expected. Images are rendering correctly. Minor tweaks to the order of fields make things easier. Thanks!

danahertzberg commented:
> Works for me :)
> There was an error when entering after uploading the media, but appears to be a Media/Core problem? This is all I could find relating to the error:
> 
> * https://www.drupal.org/forum/general/general-discussion/2020-10-20/the-specified-ajax-callback-is-empty-or-not-callable
> * https://www.drupal.org/project/drupal/issues/2866330
>   Error:
>   `Symfony\Component\HttpKernel\Exception\HttpException: The specified #ajax callback is empty or not callable. in Drupal\Core\Form\FormAjaxResponseBuilder->buildResponse() (line 67 of /src/az-quickstart-scaffolding/web/core/lib/Drupal/Core/Form/FormAjaxResponseBuilder.php).`

Follow up ticket: https://github.com/az-digital/az_quickstart/issues/638
